Action, Handlerの実装

otofu-square opened the pull request:
PRの目的：
今後の実装の方向性に間違いがないか確認をお願いします。

実装内容：
"ruboty zihou hogehoge"と打つことでhogehogeの内容に従ったメッセージ内容を出力するよう以下の通り実装
- /lib/ruboty/handlers/sebastian.rbにコマンドを実装(ruboty zihou hogehoge)
- /lib/ruboty/actions/zihou.rbにzihouコマンドのアクションを実装
  - Zihou#call : zihouコマンドで受け取ったhogehogeをZihou#zihouに渡しメッセージを受け取る
  - Zihou#zihou : 受け取ったhogehogeに対応するメッセージを返す
  - Zihou#sheduler_today : 朝の挨拶のメッセージを生成する
  - Zihou#scheduler : 予定のメッセージを生成する（サイボウズに依存する機能のため移植不要？）
  - Zihou#before_period : 定時前のメッセージを生成する
  - Zihou#period : 業務時間終了のメッセージを生成する
- setting.ymlを読み込むためyamlをrequireしたのでgemspecに依存関係を追記

まだ実装していない箇所：
各メッセージ生成メソッド内のロジック

特に見てほしい箇所：
/actions/zihou.rbの処理について、このままメソッドの中身を記述していく方向で問題がないか確認頂ければと思います。

Commit:
Merge branch 'master' into feature-Zihou

Commit:
時報機能の仮実装
 - /lib/ruboty/にhandlersディレクトリ、actionsディレクトリを追加
 - handlersディレクトリ内にsebastian.rbを実装
   - zihouコマンドを実装('ruboty zihou'で実行)
 - actionsディレクトリ内にzihou.rbを実装
   - zihouコマンドのアクションを実装(zihouコマンドを実行すると'test'と返す)
 - /lib/ruboty/sebastian.rbに上記追加ファイルをrequireするよう記述
 - 発言内容を格納するsetting.ymlを追加

Commit:
時報機能の仮実装
 - /lib/ruboty/handlers/sebastian.rbにコマンドを実装(ruboty zihou hogehoge)
 - /lib/ruboty/actions/zihou.rbにzihouコマンドのアクションを実装
   - Zihou#call : zihouコマンドで受け取ったhogehogeをZihou#zihouに渡しメッセージを受け取る
   - Zihou#zihou : 受け取ったhogehogeに対応するメッセージを返す
   - Zihou#sheduler_today : 朝の挨拶のメッセージを生成する
   - Zihou#scheduler : 予定のメッセージを生成する（）
   - Zihou#before_period : 定時前のメッセージを生成する
   - Zihou#period : 業務時間終了のメッセージを生成する
 - setting.ymlを読み込むためyamlをrequireしたのでgemspecに依存関係を追記

mgi166 commented:
1. bundle に失敗します。ruboty-sebastian.gemspec を直す必要があるかと。

``` zsh
% bundle
Fetching gem metadata from https://rubygems.org/...........
Fetching version metadata from https://rubygems.org/..
Resolving dependencies...
Bundler could not find compatible versions for gem "yaml":
  In Gemfile:
    ruboty-sebastian (>= 0) ruby depends on
      yaml (>= 0) ruby
Could not find gem 'yaml (>= 0) ruby', which is required by gem 'ruboty-sebastian (>= 0) ruby', in any of the sources.
```
1. class 名の `zihou` も英語の名詞形が好ましい
2. メソッド名も動詞が好ましい

後、setting.yml 周りのロジックは自作の予定でしょうか？
詳細を含め、口頭で確認させてください。

kano-e commented on lib/ruboty/setting.yml:
```diff
@@ -0,0 +1,125 @@
+      power_failure:
+      - "次の出社までの間に停電の予定がございます。お帰りの際に電源の状態をご確認ください。"
+
+abyss:
```
あの、ですね。
 `abyss` はもう必要ないと思うのですよ (挙動不審)

otofu-square replied:
気になっていました笑
sebastianのコードを見ると2014年1月16日にabyssモードが発動するようになっています

kano-e replied:
詳細は → https://github.com/feedforce/sebastian/pull/1#discussion-diff-21212904
ただのお遊びです :see_no_evil:

kano-e commented:
> Zihou#scheduler : 予定のメッセージを生成する（サイボウズに依存する機能のため移植不要？）

そうですね、Sebastianのスケジューラはサイボウズ使ってた頃の名残なので、今となっては移植できません。
なので、

> Zihou#sheduler_today : 朝の挨拶のメッセージを生成する

こちらも適切に名付けていただけると嬉しいです。
(この名前は、朝一に当日の予定を読み上げていた頃の名残です)

ちょっと話ずれますが、これ、ふわっとした印象だと "時報" っていうより "挨拶" って感じですね (感想)

mgi166 commented on lib/ruboty/setting.yml:
```diff
@@ -0,0 +1,125 @@
+defaults:
+  targets:
+    slack: &slack
```
これも今は不要かと。

otofu-square replied:
@kano-e 
レビューありがとうございます。
schedulerの件、教えていただきありがとうございます。
おっしゃる通り命名の件に関しては茂木さんからも指摘を受けているので、適切な名前に変更したいと思います。

@mgi166 
レビューありがとうございます。
SlackのBot用の記述だと思われるのでおっしゃる通りだと思います。

mgi166 commented on lib/ruboty/actions/zihou.rb:
```diff
@@ -0,0 +1,53 @@
+
+      private
+      # setting.ymlから表示するメッセージを取得
+      def zihou
```
このメソッドでやっていることは `Handler` でやる事かも、という話を口頭でしました。

Commit:
Handler, Actionの実装
 - Actionで処理を分岐させるべきではないとの指摘を受けたので、handlerでアクションを振り分ける
   - zihou.rbを削除し, handlerとactionの見直し
   - /lib/ruboty/actions/にmorning_greeting.rb, greeting_before_close.rb, closing_greeting.rbの追加(中身は未実装)
 - setting.ymlのような設定ファイルはルートディレクトリに置くべきとの指摘を受けたので場所を変更
 - 今回の変更に伴う/lib/ruboty/sebastian.rbの修正(requireするファイルの変更)
 - gemspecに標準ライブラリを依存関係に加えてしまったためbundlerが動かなかったのでyamlの記述を削除

Commit:
Handler, Actionの実装
 - クラス名、メソッド名を見直し
 - setting.yml -> settings.ymlにリネームし、settings.ymlから不要な情報削除

Commit:
- /lib/ruboty/sebastian.rbのrequireの箇所をクラス名、メソッド名見直しに対応

Commit:
- Settings.ymlから不要な記述を削除

mgi166 commented on lib/ruboty/handlers/sebastian.rb:
```diff
@@ -0,0 +1,23 @@
+      # コマンドのパターン
+      on /greet morning/, name: 'greet_at_morning', description: 'Output greeting message in the morning'
+      on /greet before close/, name: 'greet_before_closing', description: 'Output greeting message before close of buissness'
+      on /greet close/, name: 'greet_at_closing', description: 'Output greeting message for close of bissness'
```
bissness がタイポ

mgi166 commented on Gemfile:
```diff
@@ -1,4 +1,5 @@
 source 'https://rubygems.org'
 
+gem 'settingslogic'
```
`settingslogic` 追加したけど、今は使えないと思います。クラスを作らないと。

mgi166 commented:
`settingslogic` は次の PR で何とかすればよいかと思います。
けど、それなら今追加しなくてもいいかもしれない。タイポ直ったらマージしますよ。

Commit:
/lib/ruboty/handlers/sebastian.rbのタイポを修正

mgi166 commented:
LGTM

yukiyan commented:
あくまで私の好みなんですけど、PRの文の見出しは大きくしたほうがいいかもしれません。
細かいことで恐縮ですが、、、
[PRの綺麗な書き方はこちらをどうぞ](https://feedforce.qiita.com/yukiyan/items/37729505e7164ad4ee8a)

PRの目的：
↓

## PRの目的

otofu-square commented:
@yukiyan 
ありがとうございます！
確かにかなり見やすくなると思うので、次回からそのようにします。
